src/main/java/it/rate/webapp/services/PermissionService.java

```diff
@@ -85,18 +68,37 @@ public boolean createPlace(Long interestId) {
 
   private boolean canRateOrCreate(Interest i) {
     AppUser user = userService.getAuthenticatedUser();
-    if (user == null) {
-      return false;
-    }
-    if (user.getServerRole().equals(ServerRole.ADMIN)) {
-      return true;
-    }
-    Optional<Role> optRole = roleRepository.findById(new RoleId(user.getId(), i.getId()));
-    if (!i.isExclusive()) {
-      return true;
+    if (isAuthenticated(user)) {
+      return (isAdmin(user)
+          || !i.isExclusive()
+          || isVoter(user, i.getId())
+          || isCreator(user, i.getId()));
     }
-    return optRole.isPresent()
-        && (optRole.get().getRoleType().equals(Role.RoleType.VOTER)
-            || optRole.get().getRoleType().equals(Role.RoleType.CREATOR));
+    return false;
+  }
+
+  public boolean hasRatedOrReviewedPlace(Place place) {
+    AppUser user = userService.getAuthenticatedUser();
+    return user != null
+        && (ratingService.existsByUserAndPlace(user, place)
+            || reviewService.findById(new ReviewId(user.getId(), place.getId())).isPresent());
+  }
+
+  private boolean isAuthenticated(AppUser user) {
+    return user != null;
+  }
+
+  private boolean isAdmin(AppUser user) {
+    return user.getServerRole().equals(ServerRole.ADMIN);
+  }
+
+  private boolean isCreator(AppUser user, Long interestId) {
+    Optional<Role> optRole = roleRepository.findById(new RoleId(user.getId(), interestId));
+    return optRole.map(role -> role.getRoleType().equals(Role.RoleType.CREATOR)).orElse(false);
+  }
+
+  private boolean isVoter(AppUser user, Long interestId) {
+    Optional<Role> optRole = roleRepository.findById(new RoleId(user.getId(), interestId));
+    return optRole.map(role -> role.getRoleType().equals(Role.RoleType.VOTER)).orElse(false);

```

Comment:
Optimization suggestion:
The calling of isVoter() method is followed by calling isCreator() every time in the code which leads to calling roleService.findById() two times in a row for the same result.
isVoter() method could be replaced with something like isVoterOrCreator() which would call the database only once to get the users role.